### PR TITLE
DEVOPS-671 Adding a new CI merging rust and dart ci

### PR DESCRIPTION
**Pull request by jorge-cr-13**
Creating a new CI with the merger of the rust and dart CI's for bors.

### CODE REVIEW

**Comment by jorge-cr-13** [comment]
I'm merging the dart CI and the rust CI, there is a new ci-ok job that will be used by bors.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,352 @@
...
+        id: filter
+        with:
+          filters: |
+            dart:
+              - '.github/workflows/dart_ci.yml'
+              - 'discovery_engine/**'
+              - 'discovery_engine_core/**'
+              - 'job_runner.sh'
+            rust:
+              - '.github/workflows/rust_ci.yml'
```
Comment (rustonaut) [inline .github/workflows/ci.yml:48]: As I understand this replaces `dart_ci.yml` and `rust_ci.yml` so they can be removed from the files which trigger job runs (and deleted, I guess).

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+              - 'discovery_engine_core/**'
+
+  # DART CI 
+  dart-format:
+    name: dart-format
+    needs: selection
+    if: ${{ needs.changes.outputs.dart == 'true' }}
```
Comment (rustonaut) [inline .github/workflows/ci.yml:55]:
```suggestion
    if: ${{ needs.selection.outputs.dart == 'true' }}
```
or do I overlooked something?

Same for the other ifs.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+    runs-on: ubuntu-20.04
+    timeout-minutes: 5
+    steps:
+      - name: CI OK Testing
+        if: ${{ needs.selection.outputs.dart == 'true' }} || ${{ needs.selection.outputs.rust == 'true' }}
+        run: |
+          if ( ${{ needs.selection.outputs.dart == 'true' }} && ${{ needs.selection.outputs.rust == 'true' }} && ${{ needs.dart-doc.result == 'success'}} && ${{ needs.cargo-ok.result == 'success'}} )
+          then
+            exit 0
+          elif ( ${{ needs.selection.outputs.dart == 'true' }} && ${{ needs.dart-doc.result == 'success'}} && ${{ needs.changes.outputs.test2 == 'false' }} )
```
Comment (rustonaut) [inline .github/workflows/ci.yml:343]: What is the `test2` output for?

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+      - cargo-test
+      - cargo-doc
+    runs-on: ubuntu-20.04
+    steps:
+      - run: exit 0
+  
+  ci-ok:
+    name: ci-ok
+    needs: [selection, dart-doc, cargo-ok]
+    if: ${{always()}}
```
Comment (rustonaut) [inline .github/workflows/ci.yml:333]: For better understanding why is this needed, instead of just not having an if clause.

Comment (rustonaut) [inline .github/workflows/ci.yml:333]: For better understanding why is this needed, instead of just not having an if clause.

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -0,0 +1,352 @@
...
+      - name: cargo doc
+        working-directory: ${{ env.RUST_WORKSPACE }}
+        run: |
+          # workaround for some code being created after rust wants to parse it
+          cargo check --quiet 2>/dev/null || :
+          cargo doc --all-features --no-deps --document-private-items
+
+    # this is an helper that needs all the real leafs of the workflow.
+    # It makes easier notify_main_failure because we only need to check
+    # for this job
```
Comment (rustonaut) [inline .github/workflows/ci.yml:317]: This can be deleted, or updated there is also no `notify_main_failure` around anymore in this script.

**Comment by rustonaut** [comment]
you might have forgotten to push given that the last commit is from 6 hours ago?

File: .github/workflows/ci.yml (hunk 4 above)
Comment (acrrd) [reply .github/workflows/ci.yml:333]: It force it to run even if some needed job was skipped.

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -0,0 +1,352 @@
...
+    name: ci-ok
+    needs: [selection, dart-doc, cargo-ok]
+    if: ${{always()}}
+    runs-on: ubuntu-20.04
+    timeout-minutes: 5
+    steps:
+      - name: CI OK Testing
+        if: ${{ needs.selection.outputs.dart == 'true' }} || ${{ needs.selection.outputs.rust == 'true' }}
+        run: |
+          if ( ${{ needs.selection.outputs.dart == 'true' }} && ${{ needs.selection.outputs.rust == 'true' }} && ${{ needs.dart-doc.result == 'success'}} && ${{ needs.cargo-ok.result == 'success'}} )
```
Comment (acrrd) [inline .github/workflows/ci.yml:340]: Can't we write all of this as:
`(outputs.dart implies dart.result == success) and (outputs.rust implies rust.result == success)`
?
instead of having multiple cases?

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -0,0 +1,352 @@
...
+      - name: Install dart
+        uses: dart-lang/setup-dart@6a218f2413a3e78e9087f638a238f6b40893203d # v1.3
+        with:
+          sdk: ${{ env.DART_VERSION }}
+
+      - name: Check formatting
+        working-directory: ${{ env.DART_WORKSPACE }}
+        run: dart format --output=none --set-exit-if-changed .
+
+  dart-analyze:
```
Comment (acrrd) [inline .github/workflows/ci.yml:71]: as in rust we can have all dart-* run in parallel. Don't know how easy it is but would be nice if `dart-test` could depend on `rust-test` if it must run.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (jorge-cr-13) [reply .github/workflows/ci.yml:48]: Yes, good point I forgot to delete them from the repository and from this path.

File: .github/workflows/ci.yml (hunk 5 above)
Comment (jorge-cr-13) [reply .github/workflows/ci.yml:317]: Yes, Andrea told me that given that we were going to use `bors` the `notifyy_main_failure` was not needed anymore.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (jorge-cr-13) [reply .github/workflows/ci.yml:53]: Yes, that is correct. The `changes` and `test` are from my testing, but I updated them to `rust` and `dart` respectively.

File: .github/workflows/ci.yml (hunk 3 above)
Comment (jorge-cr-13) [reply .github/workflows/ci.yml]: The changes and test are from my testing, but I updated them to rust and dart respectively.

File: .github/workflows/ci.yml (hunk 4 above)
Comment (jorge-cr-13) [reply .github/workflows/ci.yml:327]: Because it needs to run after `selection`, `dart-doc` and `cargo-ok` have run, but if any has been skipped then anyways run.

File: .github/workflows/ci.yml (hunk 6 above)
Comment (jorge-cr-13) [reply .github/workflows/ci.yml:334]: Ok, I will try.

Comment (jorge-cr-13) [reply .github/workflows/ci.yml]: I made some changes, let me know if that was what you were referring to.

File: .github/workflows/ci.yml (hunk 7 above)
Comment (jorge-cr-13) [reply .github/workflows/ci.yml:66]: I'm going to create a new task for this.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (acrrd) [reply .github/workflows/ci.yml:50]: How can it work if we are looking for a key in `needs` that we don't specify?

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -41,7 +65,8 @@ jobs:
 
   dart-analyze:
     name: dart-analyze
-    needs: dart-format
+    needs: [dart-format, selection]
```
Comment (acrrd) [inline .github/workflows/ci.yml:68]: We discussed about running them in parallel. Will be implemented in another PR?

File: .github/workflows/ci.yml [hunk 9]
```diff
@@ -189,40 +216,123 @@ jobs:
...
+  ci-ok:
+    name: ci-ok
+    needs: [selection, dart-doc, cargo-ok]
+    if: ${{always()}}
+    runs-on: ubuntu-20.04
+    timeout-minutes: 5
+    steps:
+      - name: CI OK Testing
+        run: |
+          if [[ ${{ needs.selection.outputs.dart == 'true' }} && ${{ needs.selection.outputs.rust == 'true' }} && ${{ needs.dart-doc.result == 'success'}} && ${{ needs.cargo-ok.result == 'success'}} ||
```
[the remark this answers is not in the document]
Comment (jorge-cr-13) [reply .github/workflows/ci.yml:330]: There is the case where only the `dart` section changes and the `rust` jobs should not run, so we just need:
```
outputs.dart == 'true' => dart-doc.result == 'success'
```
My understanding is that the `dart` jobs have to run all the time, right?

[the remark this answers is not in the document]
Comment (acrrd) [reply .github/workflows/ci.yml:330]: dart-test will run in both cases yes. but then could be worth to move it out from `dart-ok`. 
We discuss this when we have the PR that run the dart process in parallel then and merge this in the mean while.

File: .github/workflows/ci.yml [hunk 10]
```diff
@@ -19,11 +17,37 @@ env:
...
+      rust: ${{ steps.filter.outputs.rust }}
+    steps:
+      # For pull requests it's not necessary to checkout the code
+      - uses: dorny/paths-filter@v2.10.2
+        id: filter
+        with:
+          filters: |
+            dart:
+              - 'discovery_engine/**'
+              - 'discovery_engine_core/**'
```
[the remark this answers is not in the document]
Comment (jorge-cr-13) [reply .github/workflows/ci.yml:42]: Spoke to @rustonaut  and he explained me that: 
`discovery_engine_core/` generates a C header which is then used by `discovery_engine` to generate dart code using a external tool (which happend to have been written in rust, but it's  in another repo). Through for testing dart we need to fully build rust.
`discovery_engine_flutter` will only have dart code? Yes (ignoring the kotlin/swift/etc. glue code for compiling it as app, in which case you also need fully build rust code, testing it also needs to have rust build as it might call functions implemented in rust during tests).

File: .github/workflows/ci.yml (hunk 9 above)
[the remark this answers is not in the document]
Comment (jorge-cr-13) [reply .github/workflows/ci.yml]: I made the changes, I think we have covered all the cases. I'll open a PR with the parallel runs.

File: .github/workflows/ci.yml (hunk 10 above)
[the remark this answers is not in the document]
Comment (acrrd) [reply .github/workflows/ci.yml:43]: Ok, we can see if we can do something better later with a cache for generated file bu is also a bit risky.

**Review by acrrd — APPROVED** [review]
